iOS cookie behavior blog (#1027)
oktadeveloper/okta-blog: _source/_posts/2022-01-13-mobile-sso.md

---
layout: blog_post
title: How to Create a Seamless Mobile SSO (Single Sign-On) Experience in iOS
author: huan-liu
by: internal-contributor
communities: [mobile]
description: "learn platform constraint on iOS platform and how to build great SSO experience around those constraints"
tags: [ios, sso, mobile-sso, mobile]
tweets:
- ""
- ""

> image: ""

Reviewer: ```suggestion
image: blog/mobile-sso/mobile-sso-social.jpeg
```

type: conversion
---

> On an iPhone, when we log in to an app, we click a login button, and a website pops up to verify our credentials. Once verified, the website then redirects back to the app, and you are logged in. This familiar Single Sign-On (SSO) pattern is frequently referred to as the *redirect flow* for authentication. The use of a web browser for auth in this example was considered [a "Best Current Practice"](https://www.rfc-editor.org/rfc/rfc8252.txt) for security and usability reasons.

Reviewer: ```suggestion
On an iPhone, when we log in to an app, we click a login button, and a website pops up to verify our credentials. Once verified, the website then redirects back to the app, and you are logged in. This familiar Single Sign-On (SSO) pattern is frequently referred to as the *redirect flow* for authentication. The use of a web browser for auth in this example is considered [a "Best Current Practice"](https://www.rfc-editor.org/rfc/rfc8252.txt) for security and usability reasons.
```

* **Problem 2** (Ambiguous UI message): If you implement logout functionality through the same flow, the prompt still says "Sign In", even though the user may have already clicked a "Logout" button. This too is confusing to the end user, as shown in the following screenshot. This problem is reported as [a bug on the AppAuth library](https://github.com/openid/AppAuth-iOS/issues/255), although it was designed as a security and privacy feature.

{% img blog/mobile-sso/ASWebAuthenticationSession-on-logout.jpeg alt:"permission prompt when logging out with ASWebAuthenticationSession" width:"300" %}{: .center-image }

In this article, I'll go into more detail on aspects of the iOS platform limitations. I'll explain why this prompt is shown and how to get around it to build a more seamless user experience. 

Specifically, this article will:

* explain the evolution of iOS over time, and how and why this prompt was introduced. 
* describe various ways you can invoke a standalone or embedded browser on iOS, and explain the cookie-sharing behaviors.
* discuss several ways to eliminate the confusing user experience issues we've identified above.

> 

Reviewer: ```suggestion

**Table of Contents**{: .hide }
* Table of Contents
{:toc}

```

Optional, this is a bit low on the page, so we might just want to skip adding it.  I think adding it earlier might break up the flow of the post while explaining the problem (including the images)

This solution was groundbreaking, because it is secure (the native app can neither peek into the embedded browser, nor alter its state) and it provides a better user experience by avoiding app switching. 

`SFSafariViewController` shares cookies with the standalone Safari browser. This gives developers a way to implementSSO (Single Sign-On) with shared cookies..

### iOS 11 changes (2017)

Apple changed `SFSafariViewController` behavior to address privacy concerns with the release of iOS 11 in 2017. `SFSafariViewController` no longer shares cookies with the standalone Safari browser. Thus, a website can no longer determine that the user on `SFSafariViewController` is the same user on Safari, even though the view controller and the browser are open on the same device. 

Apple understood that  the `SFSafariViewController` behavior change would break SSO, since single sign-on relies on the ability to share cookies. Instead, Apple had to introduce `SFAuthenticationSession` as a workaround, where `SFAuthenticationSession` would share a persistent cookie with Safari. 

> ### Iterative changes in iOS12 (2018)

Reviewer 2: ```suggestion
### Iterative changes in iOS 12 (2018)
```
spacing

A few third-party websites extend Apple's documentation to highlight that persistent cookies are shared between an embedded browser and Safari. Unfortunately, it is rarely known that session cookies can also be shared between different embedded browsers. If your application just needs SSO between your mobile apps, you do not have to use a persistent cookie; a session cookie is sufficient. 

The following table summarizes the complete sharing behavior in iOS 11 or later. I've omitted `SFAuthenticationSession` for brevity because it is deprecated, but it behaves the same as `ASWebAuthenticationSession`. I've also omitted the `prefersEphemeralWebBrowserSession` option for `ASWebAuthenticationSession` because when it is set, cookies are not shared at all. 
 
<div markdown="1" class="scrollable">

|   | APP1 + SFSafariViewController | App1 + ASWebAuthenticationSession | App2 + SFSafariViewController | App2 + ASWebAuthenticationSession | Safari | Other browsers (e.g., Chrome) |
|-----|-----|-----|-----|-----|-----|-----|
| Session cookie    |   | Yes |   |  Yes |     |   |
| Persistent cookie |   | Yes |   |  Yes | Yes |   | 

> </div>

Reviewer: @mraible  can you take a quick look at this, it's similar to what you did in #994

@drhuanliu it looks like we can shrink this table down a bit by removing the duplicated App 1 and App 2 (as those values are the same, maybe something like:

|   |SFSafariViewController | ASWebAuthenticationSession | Safari | Other browsers |
|-----|-----|-----|-----|-----|
| Session cookie    | ❌ | ✅ |  ❌   | ❌ |
| Persistent cookie | ❌ | ✅ | ✅ | ❌ | 

This may also help reduce the need for the scrolling (unless browsing on mobile)

Author: Cannot remove app1 vs. app2, it is showing cookie can be shared between two different apps. But I like the red/green checkmarks. Maybe I can define short acronyms for SFSafariViewController and ASWebAuthenticationSession so that the heading does not go too wide.

Reviewer: Maybe tweak the column title to be something like "Multiple ASWebAuthenticationSession" ?

Author: "Multiple ASWebAuthenticationSession" is confusing, as it may mean multiple from the same app. Trying a shorthanded version....

In the following iOS cookie behavior demo video, you can see how the session cookie and the persistent cookie are shared. The session cookie sharing behavior is not intuitive. Even when App1 is closed (which should clear all session cookies), opening App 2 would make it possible to see App1's session cookie. 

{% youtube mDytgEU0bO8 %}

It is worth noting that in iOS 14 and later you can specify another browser, such as Chrome, as the default. This does not affect the sharing behavior. `SFSafariViewController` and `ASWebAuthenticationSession` always use Safari under the hood, so they will never share cookies with browsers other than Safari, even if they are set as the default. The term *system browser*, often used in online articles, is synonymous with Safari. 

## Solution: how to remove the extra prompt

Now that we've reviewed the evolution of iOS browser behavior , and explored the rationale for the changes, let's look at solutions to improve the user experience. 

> ### Solution to problem 2

Reviewer: This seems out of order,  "Solution to problem 2" comes before "Solution to problem 1"

Author: Solution to problem 1 will also solve problem 2 as a by product. The reverse sequence is to go with smaller scope / simpler solution first.

Reviewer: AHH, I see! ❤️

Solving problem 2 only – eliminating the ambiguity of the prompt message – is straightforward: Use the browser for sign-in only; do not use the browser to sign out. You can sign out your application directly by revoking the access token and the refresh token. Okta provides a [revoke API](https://developer.okta.com/docs/guides/revoke-tokens/main/) that you can call directly. 

> Revoking a token works as a solution  if you are ok with signing the user out of the native app only. The user may still have a login session in the browser, and if the native app wants to login again, the browser will not ask the user for credentials before granting an access token. 

Reviewer: ```suggestion
Revoking a token works as a solution if you are okay with signing the user out of the native app only. The user may still have a login session in the browser, and if the native app wants to login again, the browser will not ask the user for credentials before granting an access token. 
```

if #available(iOS 13.0, *) {
    configuration?.noSSO = true
}
```

Under the hook, the `noSSO` option sets the `prefersEphemeralWebBrowserSession` flag. Note that this flag is only available in iOS 13 and above. 

Second, if you desire to support older iOS versions, you could use `SFSafariViewController` as the browser to present the login session. The following demonstrates how to launch `SFSafariViewController` if you are using the [AppAuth iOS](https://github.com/openid/AppAuth-iOS) library. 

AppAuth iOS supports a concept of *external user agent*, where you can use any browser to present the sign-in webpage. By default, if you invoke `OIDAuthState.authStateByPresentingAuthorizationRequest:presentingViewController:callback:`, it will invoke a default user agent. But you can plug in any external user agent, as long as it conforms to the `OIDExternalUserAgent` protocol. There is already an implementation of using [`SFSafariViewController` as an external agent](https://gist.github.com/ugenlik/2a543f351e9b9425800b48266760dc85). You can download it and plug it into your project. Then, when you invoke the browser, call 

> `OIDAuthState.authStateByPresentingAuthorizationRequest:presentingViewController:callback:`, and pass the external user agent `OIDExternalUserAgentIOSSafariViewController`, as shown in the following snippet. 

Reviewer: These single backtick blocks are too long and get chopped and cut off on smaller screens.

Consider rewording to wrap in a triple backtick, or otherwise introduce the fqn in the example below?
Or maybe reword to something like:

When the `OIDAuthState` callback is invoked...

(i know it's a bit of a hackie work around)

Author: Please review the new version breaking the long method into triple quote.

```swift
let appDelegate = UIApplication.shared.delegate as! AppDelegate
let externalAgent = OIDExternalUserAgentIOSSafariViewController(presentingViewController: self)

appDelegate.currentAuthorizationFlow =
        OIDAuthState.authState(byPresenting: request, externalUserAgent: externalAgent) { authState, error in
    if let authState = authState {

>         self.authState = authState
>         print("Got authorization tokens. Access token: " +
>                   "\(authState.lastTokenResponse?.accessToken ?? "nil")")

Reviewer: ```suggestion
        self.authState = authState
        print("Got authorization tokens. Access token: " +
              "\(authState.lastTokenResponse?.accessToken ?? "nil")")
```

---
layout: blog_post
title: How to Create a Seamless Mobile SSO (Single Sign-On) Experience in iOS
author: huan-liu
by: internal-contributor
communities: [mobile]

> description: "learn platform constraint on iOS platform and how to build great SSO experience around those constraints"

Reviewer 2: ```suggestion
description: "Understanding platform-level constraints is a prerequisite for developing great mobile experiences. This article explores browser options available on iOS, how they've evolved, and their cookie-sharing behaviors."
```

tags: [ios, sso, mobile-sso, mobile]
tweets:

> - ""

Reviewer 2: ```suggestion
- "Everything you need to know to create a seamless SSO experience on iOS. Plus a brief history of iOS authentication 📲 "
```

> - ""

Reviewer 2: ```suggestion
- "A brief history of the iOS SSO experience and two solutions to reduce UX confusion"
```

image: blog/mobile-sso/mobile-sso-social.jpeg

> type: conversion

Reviewer 2: ```suggestion
type: awareness
```

---

On an iPhone, when we log in to an app, we click a login button, and a website pops up to verify our credentials. Once verified, the website then redirects back to the app, and you are logged in. This familiar Single Sign-On (SSO) pattern is frequently referred to as the *redirect flow* for authentication. The use of a web browser for auth in this example was considered [a "Best Current Practice"](https://www.rfc-editor.org/rfc/rfc8252.txt) for security and usability reasons.

However, in 2017, a new prompt appeared in the login flow, before you were taken to the website. The following screenshot came from an iOS app preparing to log you in with Facebook. The interface prompt informs you that a specific app, the Yelp app in this example, wants to use a specific website to sign you in, and it warns you about information sharing. 

{% img blog/mobile-sso/ASWebAuthenticationSession-prompt-from-yelp.jpeg alt:"app login prompt when authenticating with ASWebAuthenticationSession" width:"300" %}{: .center-image }

There are a couple of problems with this prompt:

> * **Problem 1** (Where did this prompt come from?) : Many people consider this a bad user experience because the prompt looks out of place, and the wording is confusing and might alarm end users. 

Reviewer 2: ```suggestion
* **Problem One** (*Where did this prompt come from?*) : Many people consider this a bad user experience because the prompt looks out of place, and the wording is confusing and might alarm end users. 
```
style guide best practice

> * **Problem One** (*Where did this prompt come from?*) : Many people consider this a bad user experience because the prompt looks out of place, and the wording is confusing and might alarm end users. 

Reviewer 2: ```suggestion
* **Problem One** (*Where did this prompt come from?* ) : Many people consider this a bad user experience because the prompt looks out of place, and the wording is confusing and might alarm end users. 
```
adding space for visual ease

> * **Problem 2** (Ambiguous UI message): If you implement logout functionality through the same flow, the prompt still says "Sign In", even though the user may have already clicked a "Logout" button. This too is confusing to the end user, as shown in the following screenshot. This problem is reported as [a bug on the AppAuth library](https://github.com/openid/AppAuth-iOS/issues/255), although it was designed as a security and privacy feature.

Reviewer 2: ```suggestion
* **Problem Two** (*Ambiguous UI message*): If you implement logout functionality through the same flow, the prompt still says "Sign In", even though the user may have already clicked a "Logout" button. This too is confusing to the end user, as shown in the following screenshot. This problem is reported as [a bug on the AppAuth library](https://github.com/openid/AppAuth-iOS/issues/255), although it was designed as a security and privacy feature.
```
style guide best practice

> * **Problem Two** (*Ambiguous UI message*): If you implement logout functionality through the same flow, the prompt still says "Sign In", even though the user may have already clicked a "Logout" button. This too is confusing to the end user, as shown in the following screenshot. This problem is reported as [a bug on the AppAuth library](https://github.com/openid/AppAuth-iOS/issues/255), although it was designed as a security and privacy feature.

Reviewer 2: ```suggestion
* **Problem Two** (*Ambiguous UI message*): If you implement logout functionality through the same flow, the prompt still says **Sign In**, even though the user may have already clicked a **Logout** button. This too is confusing to the end user, as shown in the following screenshot. This problem is reported as [a bug on the AppAuth library](https://github.com/openid/AppAuth-iOS/issues/255), although it was designed as a security and privacy feature.
```
adding space for readability
adding bolding on Interaction elements

{% img blog/mobile-sso/ASWebAuthenticationSession-on-logout.jpeg alt:"permission prompt when logging out with ASWebAuthenticationSession" width:"300" %}{: .center-image }

In this article, I'll go into more detail on aspects of the iOS platform limitations. I'll explain why this prompt is shown and how to get around it to build a more seamless user experience. 

Specifically, this article will:

* explain the evolution of iOS over time, and how and why this prompt was introduced. 
* describe various ways you can invoke a standalone or embedded browser on iOS, and explain the cookie-sharing behaviors.
* discuss several ways to eliminate the confusing user experience issues we've identified above.

## A brief history of iOS evolution

The browser options available for authentication  on an iOS device have changed significantly over the years, as Apple continues to add capabilities to enhance user privacy.  To understand the behavior of the browser options provided on iOS, some historical context is  helpful. 

### Authentication prior to iOS 9

Before the release of iOS 9 in 2015,  there were only two ways to authenticate through a web browser, and neither one was optimal. 

> * **`UIWebView`**. This approach shows web content in a `UIView`. The mobile app can intercept interactions with the `UIView`, hence it is not secure.

Reviewer 2: ```suggestion
* **`UIWebView`**. This approach shows web content in a `UIView`. The mobile app can intercept interactions with the `UIView`, hence it is not secure.
```
spacing

> * **`UIWebView`**. This approach shows web content in a `UIView`. The mobile app can intercept interactions with the `UIView`, hence it is not secure.

Reviewer 2: ```suggestion
* **`UIWebView`**. This approach shows web content in a `UIView`. The mobile app can intercept interactions with the `UIView`,  hence it is not secure.
```
added space

> * **Redirect to an external browser**. The mobile app can open a webpage in a separate browser app, and the browser app can redirect back to the mobile app after authentication. However,app switching is not a good experience for mobile users. In addition, the redirect back can fail if another app has registered the same URL scheme first.

Reviewer 2: ```suggestion
* **Redirect to an external browser**. The mobile app can open a webpage in a separate browser app, and the browser app can redirect back to the mobile app after authentication. However, app switching is not a good experience for mobile users. In addition, the redirect back can fail if another app has registered the same URL scheme first.
```
spacing

`SFSafariViewController` shares cookies with the standalone Safari browser. This gives developers a way to implementSSO (Single Sign-On) with shared cookies..

### iOS 11 changes (2017)

Apple changed `SFSafariViewController` behavior to address privacy concerns with the release of iOS 11 in 2017. `SFSafariViewController` no longer shares cookies with the standalone Safari browser. Thus, a website can no longer determine that the user on `SFSafariViewController` is the same user on Safari, even though the view controller and the browser are open on the same device. 

Apple understood that  the `SFSafariViewController` behavior change would break SSO, since single sign-on relies on the ability to share cookies. Instead, Apple had to introduce `SFAuthenticationSession` as a workaround, where `SFAuthenticationSession` would share a persistent cookie with Safari. 

### Iterative changes in iOS 12 (2018)

In 2018, Apple deprecated `SFAuthenticationSession`, but introduced [`ASWebAuthenticationSession`](https://developer.apple.com/documentation/authenticationservices/aswebauthenticationsession) as a revised solution. 

> Both `SFAuthenticationSession` and `ASWebAuthenticationSession` are designed specifically for OAuth 2.0 authentication, not for showing general web content. To avoid abuse, when `SFAuthenticationSession` or `ASWebAuthenticationSession` are used, Apple always displays the prompt we saw earlier., The prompt is designed explicitly for user sign-in, and indicates that cookies are shared. 

Reviewer 2: ```suggestion
Both `SFAuthenticationSession` and `ASWebAuthenticationSession` are designed specifically for OAuth 2.0 authentication, not for showing general web content. To avoid abuse, when `SFAuthenticationSession` or `ASWebAuthenticationSession` are used, Apple always displays the prompt we saw earlier. The prompt is designed explicitly for user sign-in, and indicates that cookies are shared. 
```
removed typo - comma

{% img blog/mobile-sso/ASWebAuthenticationSession-prompt.jpeg alt:"permission prompt about information sharing" width:"500" %}{: .center-image }

> Apple does not know whether an `ASWebAuthenticationSession` is invoked  for sign in, sign out, or general web browsing. This is why the prompt text is generic. It only states that your app is trying to *Sign In*, regardless of the actual use case, which results in the ambiguity described earlier in  **Problem 2**. 

Reviewer 2: ```suggestion
Apple does not know whether an `ASWebAuthenticationSession` is invoked  for sign in, sign out, or general web browsing. This is why the prompt text is generic. It only states that your app is trying to **Sign In**, regardless of the actual use case, which results in the ambiguity described earlier in  **Problem 2**. 
```
using bold for referring to Interaction/interface elements (style guide)

### iOS 13 changes (2019)

> In 2019, Apple introduced `prefersEphemeralWebBrowserSession` as an option for `ASWebAuthenticationSession`. If this option is set to true, `ASWebAuthenticationSession` does not show the prompt above, and as a consequence, it does not share cookies with Safari. This gives developers a choice. Either they gain  a better user experience without no confusing prompt and no SSO, or they get  SSO, along with the annoying prompt. 

Reviewer 2: ```suggestion
In 2019, Apple introduced `prefersEphemeralWebBrowserSession` as an option for `ASWebAuthenticationSession`. If this option is set to true, `ASWebAuthenticationSession` does not show the prompt above, and as a consequence, it does not share cookies with Safari. This gives developers a choice. Either they gain  a better user experience with no confusing prompt and no SSO, or they get  SSO, along with the annoying prompt. 
```
typo

A few third-party websites extend Apple's documentation to highlight that persistent cookies are shared between an embedded browser and Safari. Unfortunately, it is rarely known that session cookies can also be shared between different embedded browsers. If your application just needs SSO between your mobile apps, you do not have to use a persistent cookie; a session cookie is sufficient. 

The following table summarizes the complete sharing behavior in iOS 11 or later. I've omitted `SFAuthenticationSession` for brevity because it is deprecated, but it behaves the same as `ASWebAuthenticationSession`. I've also omitted the `prefersEphemeralWebBrowserSession` option for `ASWebAuthenticationSession` because when it is set, cookies are not shared at all. 
I also use SVC as shorthand for `SFSafariViewController` and WAS as shorthand for `ASWebAuthenticationSession`.
 
<div markdown="1" class="scrollable">

> |   | APP1 + SVC | App1 + WAS | App2 + SVC | App2 + WAS | Safari | Other browsers (e.g., Chrome) |

Reviewer 2: ```suggestion
|   | APP1 + SVC | App1 + WAS | App2 + SVC | App2 + WAS | Safari | Other browsers (ex.: Chrome) |
```
style guide

|-----|-----|-----|-----|-----|-----|-----|
| Session cookie    | ❌ | ✅ | ❌ | ✅ | ❌ | ❌ |
| Persistent cookie | ❌ | ✅ | ❌ | ✅ | ✅ | ❌ | 

</div>

> In the following iOS cookie behavior demo video, you can see how the session cookie and the persistent cookie are shared. The session cookie sharing behavior is not intuitive. Even when App1 is closed (which should clear all session cookies), opening App 2 would make it possible to see App1's session cookie. 

Reviewer 2: ```suggestion
In the following iOS cookie behavior demo video, you can see how the session cookie and the persistent cookie are shared. The session cookie sharing behavior is not intuitive. Even when App1 is closed (which should clear all session cookies), opening App2 would make it possible to see App1's session cookie. 
```
spacing - consistency

{% youtube mDytgEU0bO8 %}

It is worth noting that in iOS 14 and later you can specify another browser, such as Chrome, as the default. This does not affect the sharing behavior. `SFSafariViewController` and `ASWebAuthenticationSession` always use Safari under the hood, so they will never share cookies with browsers other than Safari, even if they are set as the default. The term *system browser*, often used in online articles, is synonymous with Safari. 

## Solution: how to remove the extra prompt

Now that we've reviewed the evolution of iOS browser behavior , and explored the rationale for the changes, let's look at solutions to improve the user experience. 

> ### Solution to problem 2

Reviewer 2: ```suggestion
### Solution to problem two
```
style guide best practice

> Now that we've reviewed the evolution of iOS browser behavior , and explored the rationale for the changes, let's look at solutions to improve the user experience. 

Reviewer 2: ```suggestion
Now that we've reviewed the evolution of iOS browser behavior, and explored the rationale for the changes, let's look at solutions to improve the user experience. 
```
spacing

> Solving problem 2 only – eliminating the ambiguity of the prompt message – is straightforward: Use the browser for sign-in only; do not use the browser to sign out. You can sign out your application directly by revoking the access token and the refresh token. Okta provides a [revoke API](https://developer.okta.com/docs/guides/revoke-tokens/main/) that you can call directly. 

Reviewer 2: ```suggestion
Solving problem two only – eliminating the ambiguity of the prompt message – is straightforward: Use the browser for sign-in only; do not use the browser to sign out. You can sign out your application directly by revoking the access token and the refresh token. Okta provides a [revoke API](https://developer.okta.com/docs/guides/revoke-tokens/main/) that you can call directly. 
```
style guide best practice

> Solving problem two only – eliminating the ambiguity of the prompt message – is straightforward: Use the browser for sign-in only; do not use the browser to sign out. You can sign out your application directly by revoking the access token and the refresh token. Okta provides a [revoke API](https://developer.okta.com/docs/guides/revoke-tokens/main/) that you can call directly. 

Reviewer 2: ```suggestion
Solving problem two only – eliminating the ambiguity of the prompt message – is straightforward: Use the browser for sign-in only; do not use the browser to sign out. You can sign out of your application directly by revoking the access token and the refresh token. Okta provides a [revoke API](https://developer.okta.com/docs/guides/revoke-tokens/main/) that you can call directly. 
```
typo

> Revoking a token works as a solution if you are okay with signing the user out of the native app only. The user may still have a login session in the browser, and if the native app wants to login again, the browser will not ask the user for credentials before granting an access token. 

Reviewer 2: ```suggestion
Revoking a token works as a solution if you are okay with signing the user out of the native app only. The user may still have a login session in the browser, and if the native app wants to log in again, the browser will not ask the user for credentials before granting an access token. 
```
typo

This is the recommended approach if your web session may be supporting many different native apps. For example, [FB logout](https://developers.facebook.com/docs/unity/reference/current/FB.Logout/) specifically recommends not to log out of the web session. 

> However, if you  require a stronger privacy and security posture, for instance for a banking app, keeping the web session alive may not be an option. 

Reviewer 2: ```suggestion
However, if you require a stronger privacy and security posture, for instance for a banking app, keeping the web session alive may not be an option. 
```
spacing

> ### Solutions to problem 1

Reviewer 2: ```suggestion
### Solutions to problem one
```
style guide best practice

> There are two potential approaches to solve problem 1.  Both solutions use browser components that do not show a prompt. Both have the drawback that no cookie can be shared, so SSO will not work. If your app requires SSO, you'll have to find a new way to share login sessions. Fortunately, Okta recently introduced [Native SSO](https://developer.okta.com/docs/guides/configure-native-sso/main/), which allows native apps to implement single sign-on without cookies. See our [blog on SSO between mobile and desktop apps](/blog/2021/11/12/native-sso) for a full example. The following code shows how to remove the prompt, and it assumes your app either does not require SSO or uses Native SSO. 

Reviewer 2: ```suggestion
There are two potential approaches to solve problem one.  Both solutions use browser components that do not show a prompt. Both have the drawback that no cookie can be shared, so SSO will not work. If your app requires SSO, you'll have to find a new way to share login sessions. Fortunately, Okta recently introduced [Native SSO](https://developer.okta.com/docs/guides/configure-native-sso/main/), which allows native apps to implement single sign-on without cookies. See our [blog on SSO between mobile and desktop apps](/blog/2021/11/12/native-sso) for a full example. The following code shows how to remove the prompt, and it assumes your app either does not require SSO or uses Native SSO. 
```

> There are two potential approaches to solve problem one.  Both solutions use browser components that do not show a prompt. Both have the drawback that no cookie can be shared, so SSO will not work. If your app requires SSO, you'll have to find a new way to share login sessions. Fortunately, Okta recently introduced [Native SSO](https://developer.okta.com/docs/guides/configure-native-sso/main/), which allows native apps to implement single sign-on without cookies. See our [blog on SSO between mobile and desktop apps](/blog/2021/11/12/native-sso) for a full example. The following code shows how to remove the prompt, and it assumes your app either does not require SSO or uses Native SSO. 

Reviewer 2: ```suggestion
There are two potential approaches to solve problem one.  Both solutions use browser components that do not show a prompt. Both have the drawback that no cookie can be shared, so SSO will not work. If your app requires SSO, you'll have to find a new way to share login sessions. Fortunately, Okta recently introduced [Native SSO](https://developer.okta.com/docs/guides/configure-native-sso/main/), which allows native apps to implement single sign-on without cookies. See our [blog post on SSO between mobile and desktop apps](/blog/2021/11/12/native-sso) for a full example. The following code shows how to remove the prompt, and it assumes your app either does not require SSO or uses Native SSO. 
```
style guide: terminology (post vs blog)

First, you can use the `prefersEphemeralWebBrowserSession` option for `ASWebAuthenticationSession`. If you are using the [Okta OIDC iOS](https://github.com/okta/okta-oidc-ios) library, you can configure the `noSSO` option as follows:

```swift
let configuration = OktaOidcConfig(with: {YourOidcConfiguration})
if #available(iOS 13.0, *) {
    configuration?.noSSO = true
}
```

> Under the hook, the `noSSO` option sets the `prefersEphemeralWebBrowserSession` flag. Note that this flag is only available in iOS 13 and above. 

Reviewer 2: ```suggestion
Under the hood, the `noSSO` option sets the `prefersEphemeralWebBrowserSession` flag. Note that this flag is only available in iOS 13 and above. 
```
typo
